### ISSUE TITLE
Bookkeeper crashes under high load (TooLongFrameException)

### ISSUE THREAD

**Issue by RaulGracia** [apache/bookkeeper#2497]
**BUG REPORT**

***Describe the bug***

We have been consistently observing (last months) that, under high load situations, Bookies (v4.9.2) may permanently crash. At first glance, the main reason for the problem seems that at some point, messages from Bookkeeper client (v4.9.2) could be malformed/corrupted, leading the Bookie to crash due to unknown op codes (`IllegalStateException`) and Netty frames being too long (`TooLongFrameException`).

***To Reproduce***

Whenever we sustain a load close to the max Bookie capacity for a sufficient amount of time, we often find this problem. We have seen it in different environments and with diverse types of drives. Note that the problems is not deterministic, so it may take some time (minutes?) to appear (e.g., depends if the Bookie is also busy with other tasks, like garbage collection), but eventually it appears.

***Expected behavior***

We expect 3 main behaviors from Bookkeeper:
- If the problem is related to some kind of data malformation/corruption, then it needs to be fixed.
- A back-pressure mechanism to prevent crashes and adjust the client's write throughput to what the server can absorb.
- In case of a malformed message (i.e., `TooLongFrameException`), we would expect the server to just discard the message and continue working as usual.

***Screenshots***
n/a

***Additional context***
This problem normally is seen in our benchmark experiments when we induce loads to Bookkeeper close to its maximum capacity. In this scenario, instead of inducing enough back-pressure to the client, what we observe is a problem as follows:
```
Nov 24 12:18:30 ip-10-0-0-119.us-west-2.compute.internal bookkeeper[17271]: 2020-11-24 12:18:30,193 - ERROR - [bookie-io-1-2:BookieRequestHandler@70] - Unhandled exception occurred in I/O thread or handler
Nov 24 12:18:30 ip-10-0-0-119.us-west-2.compute.internal bookkeeper[17271]: java.lang.IllegalStateException: Received unknown request op code = 84
Nov 24 12:18:30 ip-10-0-0-119.us-west-2.compute.internal bookkeeper[17271]: at org.apache.bookkeeper.proto.BookieProtoEncoding$RequestEnDeCoderPreV3.decode(BookieProtoEncoding.java:194)
Nov 24 12:18:30 ip-10-0-0-119.us-west-2.compute.internal bookkeeper[17271]: at org.apache.bookkeeper.proto.BookieProtoEncoding$RequestDecoder.channelRead(BookieProtoEncoding.java:436)
...
Nov 24 12:18:30 ip-10-0-0-119.us-west-2.compute.internal bookkeeper[17271]: 2020-11-24 12:18:30,196 - ERROR - [bookie-io-1-2:BookieRequestHandler@70] - Unhandled exception occurred in I/O thread or handler
Nov 24 12:18:30 ip-10-0-0-119.us-west-2.compute.internal bookkeeper[17271]: io.netty.handler.codec.TooLongFrameException: Adjusted frame length exceeds 5242880: 1556335546 - discarded
Nov 24 12:18:30 ip-10-0-0-119.us-west-2.compute.internal bookkeeper[17271]: at io.netty.handler.codec.LengthFieldBasedFrameDecoder.fail(LengthFieldBasedFrameDecoder.java:522)
Nov 24 12:18:30 ip-10-0-0-119.us-west-2.compute.internal bookkeeper[17271]: at io.netty.handler.codec.LengthFieldBasedFrameDecoder.failIfNecessary(LengthFieldBasedFrameDecoder.java:500)
```
From this error, we extract the following observations:
- The `TooLongFrameException` is indicating that we are sending a message that is too large compared to the `nettyMaxFrameSize` configured in the server. However, there is no way in which we are sending a message of 1.5GB in size. The events we are writing to the Bookkeeper client are strictly smaller than 1MB.
- The problem seems related to a corruption of the messaged already sent to the Bookkeeper client, given that we also see a problem like `java.lang.IllegalStateException: Received unknown request op code = 84`. This means that the header itself of the message has been corrupted, and I don't know how a client application could modify that.
- Another aspect that points to a problem in the Bookkeeper client (or maybe its interactions with Netty) is that the 3 Bookies used in this experiment got the same exception at the same time, so in my view this discards that the problem is server related.

I'm attaching the Bookkeeper logs of this reproduction of the problem.
[bookkeeper-10.0.0.119.log.gz](https://github.com/apache/bookkeeper/files/5595823/bookkeeper-10.0.0.119.log.gz)
[bookkeeper-10.0.0.240.log.gz](https://github.com/apache/bookkeeper/files/5595824/bookkeeper-10.0.0.240.log.gz)
[bookkeeper-10.0.0.251.log.gz](https://github.com/apache/bookkeeper/files/5595825/bookkeeper-10.0.0.251.log.gz)

**Comment by fpj** [comment]
This issue seems to be related to the issue described in this PR:

https://github.com/apache/bookkeeper/pull/1240

Looking at the description of the PR, it doesn't look like we figured the source of the corruption. The next PR that touches the associated code in `BookieProtoEnconding` is this one:

https://github.com/apache/bookkeeper/pull/1286

**Comment by eolivelli** [comment]
@fpj regarding #1286 the problem was with JDK9+ and the fix was not really a "fix", I never found the root cause.

I am still finding double releases with clients that use v2 protocol (like Pulsar) and the client issues some command not supported by v2 (like write explicit LAC)

@RaulGracia are you using v2 protocol ? which JDK version ?

**Comment by RaulGracia** [comment]
Thanks @eolivelli, regarding your questions:
- No, we are using V3 protocol version (this is the client configuration default and we do not change it).
- We have seen this problem with both JDK8 and JDK11.

I will try with newer Bookkeeper version, but due to the nature of the failure, we suspect that it could be related to the Bookkeeper client, as it may send malformed/corrupted messages to the Bookies (i.e., in the logs posted before, the 3 Bookies crashed at the same time due to the same reason, which means that the message sent to all the Bookies was corrupted in the same way).

**Comment by eolivelli** [comment]
Do you have ackquorumsize < writequorumsize ?

**Comment by RaulGracia** [comment]
@eolivelli yes, in these experiments we use that configuration.

**Comment by eolivelli** [comment]
oh, IIRC @karanmehta93 (or someone from SF) experienced the same problem,
at that time the problem was that the application reused the same byte[] (or ByteBuf) that was passed to the BK LedgerHandle.
Basically once you pass your byte[] or ByteBuf to BK you cannot use it anymore.

**Comment by RaulGracia** [comment]
thanks @eolivelli, this is indeed a potential reason for this problem to happen: if the client is using a byte[] and it is concurrently modified by the application, it may lead to this kind of issues. However, I remember discussions about the possibility this happens with @andreipaduroiu and I remember that he was quite sure that such byte[] are not modified once delivered to the Bookkeeper client. @andreipaduroiu knows much better than me how this is guaranteed in Pravega, so perhaps he can elaborate on that.

**Comment by merlimat** [comment]
> A back-pressure mechanism to prevent crashes and adjust the client's write throughput to what the server can absorb.

The problem is in that there is no back-pressure in the client when `ack < write`. 

With `3-3-2`, for example, if one bookie is slower the client will retain the original entry until it's acked by the 3 bookies, making the amount of memory required in bk client to be unbound.

**Comment by andreipaduroiu** [comment]
We generate a CompositeByteBuf for every write using unique heap buffers backed by byte arrays. We never reuse the buffers or underlying arrays, nor do we touch those buffers again after creating them and sending them to the BK client.

**Comment by eolivelli** [comment]
It would be good to see this work committed to our community repo
https://github.com/apache/bookkeeper/pull/2182
Basically it is a checksum check on the bookie side, this way if the client sends garbage we can detect it immediately.

**Comment by eolivelli** [comment]
@RaulGracia if you run the test with 3-3-3 (or WQ =AQ) do you see the same problem?

**Comment by karanmehta93** [comment]
@merlimat We do timeout individual bookie operations as well, right? These would probably result in a delayed write failure and an ensemble change for a later write. (It is still possible to overload the client if the ingest rate is too high)

**Comment by karanmehta93** [comment]
@andreipaduroiu Can you describe your bytebuf in a bit more depth? Can you also provide more information as to which Bk client API is used?
The current contract is that all buffers passed to the client will automatically be released once write is completed.

**Comment by RaulGracia** [comment]
I have been testing a change in the last couple of days that seems to have fixed the problem: As the problem looks client-related, I have updated the Bookkeeper client form `4.9.2` to `4.12.0` in Pravega to execute benchmark experiments (Bookies are still `4.9.2`). After doing this change, I have not been able to reproduce again the `TooLongFrameException` at the server side in the last 2 days (normally, this problem is likely to occur under high enough load). That is, while we reach saturation scenarios in both cases, after upgrading the Bookkeeper client things come back to normal when the load is reduced without inducing Bookies to crash.

While we expected that a client upgrade could solve the issue, @fpj and myself have been trying to understand what has been the actual change that made things better. After looking into the release notes of Bookkeeper versions, we realized that since Bookkeeper `4.10.0`, the Netty version has been upgraded from `4.1.31.Final` to `4.1.32.Final`: https://github.com/apache/bookkeeper/pull/1934

Looking into the release of Netty `4.1.32.Final`, we see the following: https://netty.io/news/2018/11/29/4-1-32-Final.html
> Fix incorrect sizing of temp byte arrays in (Unsafe)ByteBufUtil

Which is related to this PR: https://github.com/netty/netty/pull/8484.

As the Bookkeeper client is heavily dependent on the classes fixed there, we thought that it is likely that the Netty upgrade has been the reason for this problem to disappear (or at least, to be much harder to reproduce as I have not been yet able to do that). In any case, it looks that newer version of Pravega will need to upgrade Bookkeeper to mitigate this problem.

I think that we can close this issue for now, unless I find a reproduction of the problem with the newer version of the Bookkeeper client.

Thanks for your feedback!

**Comment by karanmehta93** [comment]
FYI @jvrao @Ghatage , something that we might also run into (under heavy load). 
TLDR, netty version should be upgraded. 
I have seen instances of this stack trace before in some first party clusters.

**Comment by Ghatage** [comment]
@karanmehta93 Thanks for bringing this up!
I think we can go only as high as `4.1.50.Final` before we start hitting other issues with dependent services.

**Comment by Ghatage** [comment]
@karanmehta93 Thanks for bringing this up!
I think we can go only as high as 4.1.50.Final before we start hitting other issues.

**Comment by fpj** [comment]
@karanmehta93 @Ghatage do you know which version of BK you're using when you hit this issue? As @RaulGracia pointed out, there are a few bug fixes in Netty 4.1.32-FINAL that might have resolved this, and if I'm not mistaken, the upgrade from `4.1.31.Final` to `4.1.32.Final` happened in `4.10.0`.

I'm asking because it would be important to know whether the upgrade to `4.1.32.Final` is sufficient for this issue or we need another upgrade.

**Comment by karanmehta93** [comment]
Looks like we are currently using `4.1.22.Final` version. I might have to do a deeper dive to check if we still see these ST's.

**Comment by shoothzj** [comment]
closed by no updates.
